src/serialize.rs

```diff
@@ -0,0 +1,284 @@
+use crate::{Prefix, XorName};
+use serde::{
+    de::{self, MapAccess, SeqAccess, Visitor},
+    ser::SerializeStruct,
+    Deserialize, Deserializer, Serialize, Serializer,
+};
+use std::fmt;
+
+impl Serialize for XorName {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        // Return string with hexadecimal representation
+        if serializer.is_human_readable() {
+            return serializer.serialize_str(&hex::encode(self.0));
+        }
+
+        // Default serialization.
+        serializer.serialize_newtype_struct("XorName", &self.0)
+    }
+}
+
+impl<'de> Deserialize<'de> for XorName {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        if deserializer.is_human_readable() {
+            struct XorNameHexStrVisitor;
+            impl<'de> Visitor<'de> for XorNameHexStrVisitor {
+                type Value = XorName;
+
+                fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                    write!(formatter, "32 byte hex string")
+                }
+
+                fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
+                where
+                    E: de::Error,
+                {
+                    let buffer = <[u8; 32] as hex::FromHex>::from_hex(s)
+                        .map_err(|e| E::custom(std::format!("hex decoding ({})", e)))?;
+                    Ok(XorName(buffer))
+                }
+            }
+            return deserializer.deserialize_str(XorNameHexStrVisitor);
+        }
+
+        struct XorNameNewtypeVisitor;
+        impl<'de> Visitor<'de> for XorNameNewtypeVisitor {
+            type Value = XorName;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                write!(formatter, "tuple struct XorName")
+            }
+
+            fn visit_newtype_struct<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+            where
+                D: Deserializer<'de>,
+            {
+                Ok(XorName(<[u8; 32] as Deserialize>::deserialize(
+                    deserializer,
+                )?))
+            }
+
+            // A method like this was generated (August 2022) by #[derive(Deserialize)], so assumed to be of use
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: SeqAccess<'de>,
+            {
+                match seq.next_element::<[u8; 32]>()? {
+                    Some(a) => Ok(XorName(a)),
+                    None => Err(de::Error::invalid_length(
+                        0,
+                        &"tuple struct XorName with 1 element",
+                    )),
+                }
+            }
+        }
+
+        deserializer.deserialize_newtype_struct("XorName", XorNameNewtypeVisitor)
```

Comment:
Can we replace this custom deserializer with:
```suggestion
        Ok(XorName(<[u8; 32] as Deserialize>::deserialize(
            deserializer,
        )?))
```

Reply:
I am not sure that works?

Reply:
Though this works:
```rust
        #[derive(Deserialize)]
        #[serde(rename = "XorName")] 
        struct XorNameDerived([u8; 32]);
        let a = <XorNameDerived as Deserialize>::deserialize(
            deserializer,
        )?;
        Ok(XorName(a.0))
```

Reply:
This actually also works to remove the other Deserialize for `Prefix`:
```rust

        #[derive(Deserialize)]
        #[serde(rename = "Prefix")]
        struct PrefixDerived {
            bit_count: u16,
            name: XorName,
        }
        let a = <PrefixDerived as Deserialize>::deserialize(deserializer)?;
        Ok(Prefix { bit_count: a.bit_count, name: a.name })
```

Don't know how this impacts performance?

Reply:
Oh, yeah, that's much nicer, I wouldn't worry about perf for such a small type.

Reply:
This is clearly correct, I wasn't so confident about the other visitor implementations